File: gris/www/recepcao/novos_associados.py
"""Lista de novos associados em tabela, com link para a ficha e para o formulário de registro.

A visão geral do funil é um kanban: boa para mover o fluxo, ruim para achar alguém que está
fora da coluna em que se procura. Esta página é a lista completa — todo mundo que está em
integração, em qualquer status, com busca e filtros — e é por ela que se chega na ficha de
registro (`/recepcao/ficha_registro`) e no formulário do responsável (`/responsavel/registro`)
de qualquer jovem, sem depender do status em que ele está.

As linhas chegam por ``listar``, e não no contexto da página: é o mesmo desenho das outras
listas do portal (`/associados/lista`, `/recepcao/pesquisa_novos_respostas`), em que filtrar e
paginar não recarrega a página inteira.
"""

import frappe
from frappe import _
from frappe.rate_limiter import rate_limit
from frappe.utils import cint, format_date, get_fullname

from gris.api.portal_access import enrich_context, user_has_access
from gris.api.recepcao import formatar_idade

no_cache = 1

CAMINHO = "/recepcao/novos_associados"

RAMOS = ["Filhotes", "Lobinho", "Escoteiro", "Sênior", "Pioneiro"]

STATUS = [
	"Novo Contato",
	"Conversa Inicial",
	"Visita Agendada",
	"Aguardar Dados",
	"Fazer Registro",
	"Acompanhamento",
	"Fila de espera",
	"Concluído",
]

TIPOS_DE_REGISTRO = ["Provisório", "Definitivo"]

# Variantes do badge Basecoat por ramo, as mesmas de visao_geral.py e fila_espera.py.
VARIANTE_POR_RAMO = {
	"Filhotes": "ramo-filhotes",
	"Lobinho": "ramo-lobinho",
	"Escoteiro": "ramo-escoteiro",
	"Sênior": "ramo-senior",
	"Pioneiro": "ramo-pioneiro",
}

POR_PAGINA = 50

CAMPOS_DA_LISTA = [
	"name",
	"nome_completo",
	"data_de_nascimento",
	"ramo",
	"status",
	"tipo_de_registro",
	"dados_para_registro_enviados",
	"registro_criado_no_paxtu",
	"responsavel_recepcao",
	"modified",
]


def _pode_ver() -> bool:
	"""Mesma regra das outras páginas de `/recepcao`: papel Recepcao (ou System Manager)."""
	return user_has_access(CAMINHO)


def _itens(valores: list[str], rotulo_vazio: str) -> list[dict]:
	return [{"value": "", "label": rotulo_vazio}] + [{"value": v, "label": v} for v in valores]


def _filtros(
	busca: str | None,
	status: str | None,
	ramo: str | None,
	tipo_de_registro: str | None,
	dados_enviados: str | None,
) -> dict:
	"""Filtros da listagem, validados contra as opções do schema.

	Os valores vêm do cliente, então nada entra em ``filters`` sem estar na lista de opções —
	um status inventado viraria um filtro silencioso que devolve lista vazia.
	"""
	filtros: dict[str, object] = {}

	if (status or "").strip() in STATUS:
		filtros["status"] = status.strip()

	if (ramo or "").strip() in RAMOS:
		filtros["ramo"] = ramo.strip()

	if (tipo_de_registro or "").strip() in TIPOS_DE_REGISTRO:
		filtros["tipo_de_registro"] = tipo_de_registro.strip()

	if (dados_enviados or "").strip() in ("sim", "nao"):
		filtros["dados_para_registro_enviados"] = 1 if dados_enviados.strip() == "sim" else 0

	busca = (busca or "").strip()
	if busca:
		filtros["nome_completo"] = ["like", f"%{busca}%"]

	return filtros


def _responsaveis_por_jovem(nomes: list[str]) -> dict[str, str]:
	"""Nome do primeiro responsável de cada jovem, em duas consultas (e não uma por linha)."""
	if not nomes:
		return {}

	vinculos = frappe.get_all(
		"Responsavel Vinculo",
		filters={"beneficiario_novo_associado": ["in", nomes]},
		fields=["beneficiario_novo_associado", "responsavel"],
		order_by="primeiro_responsavel desc, creation asc",
	)

	primeiro_por_jovem: dict[str, str] = {}
	for vinculo in vinculos:
		jovem = vinculo.beneficiario_novo_associado
		if vinculo.responsavel and jovem not in primeiro_por_jovem:
			primeiro_por_jovem[jovem] = vinculo.responsavel

	if not primeiro_por_jovem:
		return {}

	nomes_de_responsavel = {
		resp.name: resp.nome_completo
		for resp in frappe.get_all(
			"Responsavel",
			filters={"name": ["in", list(set(primeiro_por_jovem.values()))]},
			fields=["name", "nome_completo"],
		)
	}

	return {
		jovem: nomes_de_responsavel.get(responsavel) or ""
		for jovem, responsavel in primeiro_por_jovem.items()
	}


def _linhas(registros: list[dict]) -> list[dict]:
	"""Linhas da tabela, já com o que a página precisa mostrar em cada coluna."""
	responsaveis = _responsaveis_por_jovem([registro.name for registro in registros])
	emails_da_recepcao = {
		registro.responsavel_recepcao for registro in registros if registro.responsavel_recepcao
	}
	nomes_da_recepcao = {email: get_fullname(email) for email in emails_da_recepcao}

	return [
		{
			"name": registro.name,
			"nome_completo": registro.nome_completo or _("(sem nome)"),
			"idade": formatar_idade(registro.data_de_nascimento) or "",
			"ramo": registro.ramo or "",
			"ramo_variante": VARIANTE_POR_RAMO.get(registro.ramo or "", "secondary"),
			"status": registro.status or "",
			"tipo_de_registro": registro.tipo_de_registro or "",
			"dados_enviados": 1 if cint(registro.dados_para_registro_enviados) else 0,
			"registro_criado_no_paxtu": 1 if cint(registro.registro_criado_no_paxtu) else 0,
			"responsavel_legal": responsaveis.get(registro.name) or "",
			"responsavel_recepcao": nomes_da_recepcao.get(registro.responsavel_recepcao) or "",
			"atualizado_em": format_date(registro.modified) if registro.modified else "",
		}
		for registro in registros
	]


@frappe.whitelist()
@rate_limit(key="recepcao-lista-novos", limit=120, seconds=60)
def listar(
	busca: str | None = None,
	status: str | None = None,
	ramo: str | None = None,
	tipo_de_registro: str | None = None,
	dados_enviados: str | None = None,
	pagina: int | str = 1,
) -> dict:
	"""Página da lista de novos associados, com os filtros aplicados."""
	if frappe.session.user == "Guest":
		frappe.throw(_("Você precisa estar logado."), frappe.PermissionError)

	if not _pode_ver():
		frappe.throw(_("Acesso permitido apenas para Recepção."), frappe.PermissionError)

	filtros = _filtros(busca, status, ramo, tipo_de_registro, dados_enviados)

	total = frappe.db.count("Novo Associado", filtros)
	ultima_pagina = max(1, -(-total // POR_PAGINA))
	# Página além do fim (o filtro mudou e a paginação ficou para trás) volta para a primeira,
	# senão a tela aparece vazia enquanto o contador diz que há registros.
	pagina = max(1, cint(pagina) or 1)
	if pagina > ultima_pagina:
		pagina = 1

	registros = frappe.get_all(
		"Novo Associado",
		filters=filtros,
		fields=CAMPOS_DA_LISTA,
		order_by="nome_completo asc",
		limit_start=(pagina - 1) * POR_PAGINA,
		limit_page_length=POR_PAGINA,
	)

	return {
		"linhas": _linhas(registros),
		"total": total,
		"pagina": pagina,
		"ultima_pagina": ultima_pagina,
		"por_pagina": POR_PAGINA,
		"com_filtro": bool(filtros),
	}


def get_context(context):
	if frappe.session.user == "Guest":
		frappe.local.flags.redirect_location = f"/login?redirect-to={CAMINHO}"
		raise frappe.Redirect

	if not _pode_ver():
		frappe.throw(_("Acesso permitido apenas para Recepção."), frappe.PermissionError)

	context.itens_status = _itens(STATUS, "Todos")
	context.itens_ramo = _itens(RAMOS, "Todos")
	context.itens_tipo_de_registro = _itens(TIPOS_DE_REGISTRO, "Todos")
	context.itens_dados_enviados = [
		{"value": "", "label": "Todos"},
		{"value": "sim", "label": "Enviados"},
		{"value": "nao", "label": "Pendentes"},
	]
	context.por_pagina = POR_PAGINA

	context.active_link = CAMINHO
	enrich_context(context, CAMINHO)

	return context

File: gris/www/recepcao/novos_associados.js
// Lista de Novos Associados (/recepcao/novos_associados)
// As linhas vêm de `gris.www.recepcao.novos_associados.listar`: filtrar e paginar troca só
// o corpo da tabela, sem recarregar a página.

(function () {
	const COLSPAN = 10;

	const escapeHtml = (valor) => frappe.utils.escape_html(valor == null ? "" : String(valor));

	const textoOuTraco = (valor) => (valor ? escapeHtml(valor) : "—");

	frappe.ready(() => {
		const form = document.getElementById("novos-filtros");
		const tabela = document.getElementById("novosAssociadosTable");
		const tbody = tabela?.querySelector("tbody");
		const wrapper = document.getElementById("novos-tabela-wrap");
		const carregando = document.getElementById("novos-loading");
		const vazio = document.getElementById("novos-empty");
		const contador = document.getElementById("novos-contador");
		const btnAnterior = document.getElementById("btn-pagina-anterior");
		const btnProxima = document.getElementById("btn-pagina-proxima");
		const btnLimpar = document.getElementById("btn-limpar");

		if (!tbody || !form) return;

		let pagina = 1;
		let ultimaPagina = 1;
		let carregandoAgora = false;

		function filtrosAtuais() {
			const dados = new FormData(form);
			return {
				busca: (dados.get("busca") || "").toString().trim(),
				status: (dados.get("status") || "").toString(),
				ramo: (dados.get("ramo") || "").toString(),
				tipo_de_registro: (dados.get("tipo_de_registro") || "").toString(),
				dados_enviados: (dados.get("dados_enviados") || "").toString(),
			};
		}

		function mostrarEstado(estado) {
			carregando.hidden = estado !== "carregando";
			wrapper.hidden = estado !== "lista";
			vazio.hidden = estado !== "vazio";
		}

		function descreverVazio(comFiltro) {
			const descricao = vazio.querySelector("p");
			if (!descricao) return;
			descricao.textContent = comFiltro
				? "Revise a busca e os filtros aplicados."
				: "Ninguém em integração no momento.";
		}

		function badgeRamo(linha) {
			if (!linha.ramo) return "—";
			return `<span class="badge badge-${escapeHtml(linha.ramo_variante)}">${escapeHtml(
				linha.ramo
			)}</span>`;
		}

		function badgeDados(linha) {
			if (linha.dados_enviados) {
				return '<span class="badge badge-dados-enviados">Enviados</span>';
			}
			return '<span class="badge-outline">Pendentes</span>';
		}

		function acoes(linha) {
			const nome = encodeURIComponent(linha.name);
			const rotuloRegistro = linha.registro_criado_no_paxtu
				? "Formulário do responsável"
				: "Preencher registro em nome do responsável";
			return `
				<div class="novos-acoes">
					<a
						href="/recepcao/ficha_registro?name=${nome}"
						class="btn-sm-outline"
						title="Abrir a ficha de registro"
					>Ficha</a>
					<a
						href="/responsavel/registro?novo_associado=${nome}"
						class="btn-sm-ghost"
						title="${escapeHtml(rotuloRegistro)}"
					>Registro</a>
				</div>`;
		}

		function renderLinhas(linhas) {
			tbody.innerHTML = linhas
				.map((linha) => {
					const nome = encodeURIComponent(linha.name);
					return `
						<tr>
							<td data-coluna="nome">
								<a href="/recepcao/ficha_registro?name=${nome}" class="novos-nome">${escapeHtml(
									linha.nome_completo
								)}</a>
							</td>
							<td>${textoOuTraco(linha.idade)}</td>
							<td>${badgeRamo(linha)}</td>
							<td>${textoOuTraco(linha.status)}</td>
							<td>${textoOuTraco(linha.tipo_de_registro)}</td>
							<td>${badgeDados(linha)}</td>
							<td>${textoOuTraco(linha.responsavel_legal)}</td>
							<td>${textoOuTraco(linha.responsavel_recepcao)}</td>
							<td>${textoOuTraco(linha.atualizado_em)}</td>
							<td>${acoes(linha)}</td>
						</tr>`;
				})
				.join("");
		}

		function renderErro(mensagem) {
			tbody.innerHTML = `
				<tr>
					<td colspan="${COLSPAN}">
						<p class="novos-erro">${escapeHtml(mensagem)}</p>
					</td>
				</tr>`;
			mostrarEstado("lista");
		}

		function atualizarRodape(resposta) {
			const total = Number(resposta.total || 0);
			const porPagina = Number(resposta.por_pagina || 0);
			if (!total) {
				contador.textContent = "Nenhum registro.";
			} else {
				const primeiro = (pagina - 1) * porPagina + 1;
				const ultimo = Math.min(total, primeiro + resposta.linhas.length - 1);
				contador.textContent = `${primeiro}–${ultimo} de ${total} ${
					total === 1 ? "pessoa" : "pessoas"
				}`;
			}
			btnAnterior.disabled = pagina <= 1;
			btnProxima.disabled = pagina >= ultimaPagina;
		}

		function carregar() {
			if (carregandoAgora) return;
			carregandoAgora = true;
			mostrarEstado("carregando");

			frappe.call({
				method: "gris.www.recepcao.novos_associados.listar",
				args: Object.assign({ pagina: pagina }, filtrosAtuais()),
				// O msgprint do Frappe não aparece neste portal: o erro é mostrado na tabela.
				silent: true,
				callback: function (r) {
					const resposta = r.message;
					if (!resposta) {
						renderErro("Não foi possível carregar a lista.");
						return;
					}
					pagina = Number(resposta.pagina || 1);
					ultimaPagina = Number(resposta.ultima_pagina || 1);
					if (resposta.linhas.length) {
						renderLinhas(resposta.linhas);
						mostrarEstado("lista");
					} else {
						tbody.innerHTML = "";
						descreverVazio(Boolean(resposta.com_filtro));
						mostrarEstado("vazio");
					}
					atualizarRodape(resposta);
				},
				error: function () {
					renderErro("Não foi possível carregar a lista. Tente novamente.");
				},
				always: function () {
					carregandoAgora = false;
				},
			});
		}

		form.addEventListener("submit", (evento) => {
			evento.preventDefault();
			pagina = 1;
			carregar();
		});

		btnLimpar?.addEventListener("click", () => {
			// O reset nativo limpa o hidden input do select, mas não o rótulo do trigger.
			window.setTimeout(() => {
				form.querySelectorAll(".select").forEach((el) => {
					el.value = "";
				});
				pagina = 1;
				carregar();
			}, 0);
		});

		btnAnterior.addEventListener("click", () => {
			if (pagina <= 1) return;
			pagina -= 1;
			carregar();
		});

		btnProxima.addEventListener("click", () => {
			if (pagina >= ultimaPagina) return;
			pagina += 1;
			carregar();
		});

		carregar();
	});
})();
